jupyterhub/spawner.py

```diff
@@ -119,50 +165,93 @@ def options_from_form(self, form_data):
+        with one parameter (the spawner instance), and should return a string fairly quickly (no blocking
+        operations please!).
+
+        Note that the spawner class' interface is not guaranteed to be exactly same across upgrades,
+        so if you are using the callable take care to verify it continues to work after upgrades!
         """
     ).tag(config=True)
-    
-    cmd = Command(['jupyterhub-singleuser'],
-        help="""The command used for starting notebooks."""
+
+    # FIXME: Add info about shell expansion here.
+    cmd = Command(
+        ['jupyterhub-singleuser'],
+        help="""
+        The command used for starting the single-user server.
+
+        You can provide this as either a list or a string. Note that this should only contain
+        the path to the script to start - extra arguments should be provided via `args`.
+
+        This is usually set if you want to start the single-user server in a different python
+        environment (with virtualenv/conda) than JupyterHub itself.
+        """
     ).tag(config=True)
-    args = List(Unicode(),
-        help="""Extra arguments to be passed to the single-user server"""
+
+    # FIXME: Add info about shell expansion here.
+    args = List(
+        Unicode(),
+        help="""
+        Extra arguments to be passed to the single-user server.
+
+        """
     ).tag(config=True)
-    
-    notebook_dir = Unicode('',
-        help="""The notebook directory for the single-user server
-        
-        `~` will be expanded to the user's home directory
-        `{username}` will be expanded to the user's username
+
+    notebook_dir = Unicode(
+        '',
+        help="""
+        Path to the notebook directory for the single-user server.
+
+        The user will see a file listing of this directory when they start, and the Notebook
+        interface itself will not allow browsing outside the contents of this directory (and its
+        subdirectories) easily.
+
+        `~` will be expanded to the home directory of the user, and {username} will be replaced
+        with the name of the user.
+
+        Note that this does *not* prevent users from accessing files outside of this path! They
+        can do so with many other means.
         """
     ).tag(config=True)
 
-    default_url = Unicode('',
-        help="""The default URL for the single-user server. 
+    # FIXME: Provide example of such an URL!
+    default_url = Unicode(
+        '',
+        help="""
+        The URL the single-user server should start in.
 
-        Can be used in conjunction with --notebook-dir=/ to enable 
-        full filesystem traversal, while preserving user's homedir as
-        landing page for notebook
+        You can set `notebook_dir` to `/` and then construct a URL here to allow people to navigate
+        the whole filesystem from their notebook, but still start in their home directory.
```

Comment:
The default URL for starting the single-user server.

Can be used in conjunction with `notebook_dir` to provide greater flexibility for filesystem navigation. For example, setting `notebook_dir = '/'` and `default_url = '/home/{username}` permits a user to view their home directory when starting the notebook interface and to navigate the entire filesystem (not just subdirectories of the home directory).